File: source/bl/invoice-book-service/src/main/java/ca/aatl/app/invoicebook/bl/ejb/TimeEntryService.java
/*
 * Copyright (c) 2019 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2019-Jan-28  GShokar         Created
 * =============================================================================
 */
package ca.aatl.app.invoicebook.bl.ejb;

import ca.aatl.app.invoicebook.data.jpa.dao.TimeEnityDao;
import ca.aatl.app.invoicebook.data.jpa.entity.TimeEntry;
import ca.aatl.app.invoicebook.exception.DataValidationException;
import java.util.Date;
import java.util.List;
import javax.ejb.EJB;
import javax.ejb.LocalBean;
import javax.ejb.Stateless;

/**
 *
 * @author GShokar
 */

@Stateless
@LocalBean
public class TimeEntryService {

    @EJB
    TimeEnityDao dao;
    
    public TimeEntry newEntity() {
        return new TimeEntry();
    }

    public TimeEntry find(String guid) {
        return dao.find(guid);
    }

    public void validate(TimeEntry entity) throws DataValidationException {
        boolean rtnValue = true;

        StringBuilder sb = new StringBuilder();

        if (entity != null) {

            if (entity.getDate() == null) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the time entry date.");
            } 
            
            if (entity.getTimeCode() == null) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Select the time code.");
            } 
            
            if (entity.getStartTime() == null) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the start time.");
            } 
            
            if (entity.getEndTime() == null) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the end time.");
            } 
            
            if (entity.getStartTime() != null 
                    && entity.getEndTime() == null
                    && entity.getStartTime().after(entity.getEndTime())) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the end time should not be before start time.");
            } 
            
        }

        if (!rtnValue) {
            throw new DataValidationException(sb.toString());
        }
    }

    public void save(TimeEntry timeEntry) throws Exception{
        
        beforeSave(timeEntry);
        
        dao.save(timeEntry);
    }

    private void beforeSave(TimeEntry entity) throws DataValidationException {
        validate(entity);

        if (entity.getId() == null) {

            entity.setAddedBy(entity.getLastUpdatedBy());
            entity.setAddedDate(entity.getLastUpdatedDate());
            entity.getGuid();
            
        }
    }

    public List<TimeEntry> list() throws Exception{
        return dao.list();
    }

    public List<TimeEntry> find(String employeeNumber, Date yearMonthDate) throws Exception{
        return dao.find(employeeNumber, yearMonthDate);
    }
}

File: source/client/web/invoice-book-web/public_html/js/aatl-ib/service/time-entry-service.js
/* 
 * Copyright (c) 2019 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2019-Jan-28  GShokar         Created
 * =============================================================================
 */

"use strict";

$aatl_ib.TimeEntryService = {

    find: function (criteria, callback) {

        $aatl_ib.ApiService.get("time-entries/find",
                criteria,
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get time entries: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let list = JSON.parse(res.data);
                        callback(list);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });
    },

    get: function (number, callback) {

        if (number === undefined || number === null) {
            callback({
                number: "",
                lastName: "",
                firstName: "",
                birthDate: "",
                address: {
                    address1: "",
                    address2: "",
                    city: "",
                    province: $aatl_ib.LookupService.defaultProvince.code,
                    postalCode: ""
                },
                contact: {
                    phone: "",
                    email: ""
                }
            });
        } else {
            $aatl_ib.ApiService.get("employee/" + number,
                    {},
                    function (res, err) {
                        if (err) {
                            callback(null, {messages: ["Faild to get employee detail: " + err]});

                        } else if (res.status === "failure") {
                            callback(null, {messages: [res.message]});
                        } else if (res.status === "success") {
                            let employee = JSON.parse(res.data);
                            callback(employee);
                        } else {
                            callback(null, {messages: ["Invalid response from server"]});
                        }
                    });
        }
    },

    save: function (timeEntry, callback) {

        let err = $aatl_ib.TimeEntryService.validate(timeEntry);

        if (err.messages.length === 0) {
            $aatl_ib.ApiService.put("time-entries",
                    timeEntry,
                    function (res, serverErr) {
                        if (serverErr) {

                            err.messages.push("Failed to save the time entry: " + serverErr);
                            callback(timeEntry, err);

                        } else if (res.status === "failure") {

                            err.messages.push(res.message);
                            callback(timeEntry, err);

                        } else if (res.status === "success") {

                            if (Array.isArray(res.warningMessages) && res.warningMessages.length > 0) {
                                err.messages = res.warningMessages;
                            }

                            callback(JSON.parse(res.data), err);
                        } else {
                            callback(timeEntry, "Invalid response from server");
                        }
                    });

            return;
        }

        callback(timeEntry, err);
    },
    list: function (callback) {

        $aatl_ib.ApiService.get("time-entries",
                {},
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get time entry list: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let list = JSON.parse(res.data);
                        callback(list);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });
    },
    validate: function (timeEntry) {

        let err = {messages: []};

        if (timeEntry) {

            if ($aatl_ib.utils.isStringEmpty(timeEntry.date)) {
                err.messages.push("Please select enter date.");
            }
            if ($aatl_ib.utils.isStringEmpty(timeEntry.timeCode.uid)) {
                err.messages.push("Please select time code.");
            }
            if ($aatl_ib.utils.isStringEmpty(timeEntry.startTime)) {
                err.messages.push("Please enter start time.");
            }
            if ($aatl_ib.utils.isStringEmpty(timeEntry.endTime)) {
                err.messages.push("Please enter end time.");
            }

        }

        return err;
    }
};


